tsdb/engine/v1/engine.go

```diff
@@ -594,11 +597,29 @@ func (c *Cursor) read() (key, value []byte) {
 		return nil, nil
 	}
 
-	// Use the buffer if it exists and there's no cache or if it is lower than the cache.
-	if c.buf.key != nil && (c.index >= len(c.cache) || bytes.Compare(c.buf.key, c.cache[c.index][0:8]) == -1) {
-		key, value = c.buf.key, c.buf.value
-		c.buf.key, c.buf.value = nil, nil
-		return
+	// Check if the buffer exists.
+	if c.buf.key != nil {
+		// Return the buffer if the cache is drained.
+		if c.index >= len(c.cache) {
+			return c.readBuf()
+		}
+
+		// Return the buffer if less than the cache.
+		cmp := bytes.Compare(c.buf.key, c.cache[c.index][0:8])
+		if cmp == -1 {
+			return c.readBuf()
+		}
+
+		// If the buffer and cache have equal keys then drain the cache.
+		if cmp == 0 {
+			for {
+				c.index++
+				if c.index >= len(c.cache) || !bytes.Equal(c.buf.key, c.cache[c.index][0:8]) {
+					break
+				}
+			}
+			return c.readBuf()
+		}
 	}
```

Comment:
Sorry, I fixed another issue in 097b025 so that the cache is read instead of the buffer.